Enables duration queries by promoting Span.duration, timestamp

codefromthecrypt opened the pull request:
A commonly requested feature is querying by duration, for example showing traces longer than 1 minute. 

While the UI supports ordering by duration, it is a game of chance whether the traces returned happen to be long or not. Even if you put a limit of 1000, you still may not find the longest trace (as your storage system may have a million traces in it).

The path proposed is top-level Span.timestamp, and Span.duration, which formerly existed in the scala and mustache models, but not in the thrift.

We could implement duration queries without changing thrifts. For example, this could be an implementation detail of each storage system. However, this limits testing to side-effects and keeps essential details in scala code, making portability harder.

Also, we have other reasons to top-level these fields:
- local spans can be smaller (#808)
- non-rpc spans can't be supported without artificially logging annotations like cs/cr
- reduce redundant span.duration logic in several scala classes and mustache templates
- Dependency graph links can cheaply add duration aggregates

Here's a convenience paste of the definitions of Span.timestamp and Span.duration

``` java
  /**
   * Microseconds from epoch of the creation of this span.
   *
   * This value should be set directly by instrumentation, using the most
   * precise value possible. For example, gettimeofday or syncing nanoTime
   * against a tick of currentTimeMillis.
   *
   * For compatibilty with instrumentation that precede this field, collectors
   * or span stores can derive this via Annotation.timestamp.
   * For example, SERVER_RECV.timestamp or CLIENT_SEND.timestamp.
   *
   * This field is optional for compatibility with old data: first-party span
   * stores are expected to support this at time of introduction.
   */
  10: optional i64 timestamp,
  /**
   * Measurement of duration in microseconds, used to support queries.
   *
   * This value should be set directly, where possible. Doing so encourages
   * precise measurement decoupled from problems of clocks, such as skew or NTP
   * updates causing time to move backwards.
   *
   * For compatibilty with instrumentation that precede this field, collectors
   * or span stores can derive this by subtracting Annotation.timestamp.
   * For example, SERVER_SEND.timestamp - SERVER_RECV.timestamp.
   *
   * If this field is persisted as unset, zipkin will continue to work, except
   * duration query support will be implementation-specific. Similarly, setting
   * this field non-atomically is implementation-specific.
   *
   * This field is i64 vs i32 to support spans longer than 35 minutes.
   */
  11: optional i64 duration
```

yurishkuro commented on zipkin-thrift/src/main/thrift/com/twitter/zipkin/zipkinCore.thrift:
```diff
@@ -204,5 +204,19 @@ struct Span {
+   * example, SERVER_SEND.timestamp - SERVER_RECV.timestamp.
+   *
+   * Note that this should be treated unsigned. i32 implies trace durations are
+   * not longer than 35.79 minutes.
```
I don't think that's a reasonable limitation. Just the other day I had a discussion with the data team where they sometimes have spans measured in hours. I would go for i64.

codefromthecrypt replied:
fair enough.

codefromthecrypt replied:
done

yurishkuro commented on zipkin-thrift/src/main/thrift/com/twitter/zipkin/zipkinCore.thrift:
```diff
@@ -204,5 +204,19 @@ struct Span {
+   * Microseconds from epoch, usually taken from the first annotation. Ex.
+   * SERVER_RECV.timestamp
+   */
+  10: optional i64 start_ts,
```
What is the benefit of capturing this explicitly instead of the usual cs/cr? It feels like it's going to complicate other code paths as they need to handle dual representation.

codefromthecrypt replied:
PS I don't see this as complicating, particularly as I've touched all the
code in zipkin. The current codebase is a mess as it has implicit logic
around things like startTs and duration. This formalizes it, making it
possible to create an un-hairy implementation.

yurishkuro commented:
> In order to support a query by duration, we need to have an index of (service name, duration) -> traceId limited by startTs. We've always had implicit Span.duration and Span.startTs fields. To ensure we can support duration queries, we need to make these explicit.

The last sentence does not follow from the requirements. Yes, we do need an index, but we don't need to capture start/duration in the collected spans, we just need a way to calculate it in the collector.  We could  use cs/cr for that, provided they reach the collector in a single submission from instrumentation.

yurishkuro commented:
These are my concerns with adding the fields to thrift: 
- it suggests to the instrumentation libs to send redundant data, since they are already sending the start/end annotations from which startTs/duration can be derived. The Zipkin story about what and how the instrumentation libs have to send is already super convoluted.
- In RPC case the start/duration can be derived either from server annotations or from client annotations, depending on where the instrumentation exists. There isn't really such a thing as a "span" in the ingestion pipeline, more like multiple slices of span. A slice + annotations is not ambiguous (although  "span name" is), but slice + start/duration definitely are. So someone need to resolve that somewhere, and I think it's better resolved using the raw data of annotations than using potentially incorrectly calculated start/duration.
- In cases when instrumentation cannot send the complete span, some other mechanism, like a Spark pipeline, can do a post-processing and calculate start/duration having seen the complete span in the storage (multiple slices stored independently). It would certainly be storage implementation dependent how such post-caculated result would be stored. In Cassandra we'd add a record to the index by duration. If other stores have a record representing a complete span, they would have to make an update to that record. Forcing Cassandra to store an update of post-calculation would be pretty bad for performance.

yurishkuro commented:
btw, what is your concern with having these two fields be implementation-specific? The span API that query service deals with can have them as def's, and storage implementations fill it. In many cases the actual calculation should be the same, e.g. if it's from annotations, so that logic can be refactored to a base class or composition. I thought right now it it's already in the shared class.

yurishkuro commented:
ok, I'll shut up, go for it. I believe in my system we may be able to calculate those fields in the collector.

codefromthecrypt commented:
Glad it might work. Just added some more detail around how I see this fitting in with local spans. PTAL, as it may make more sense within this context https://github.com/openzipkin/zipkin/issues/808

codefromthecrypt commented:
Re-summarized in the thrift definition and WIP description. I removed my related redundant comments.

codefromthecrypt commented:
Note I plan to start on this today, as soon as the supporting changes are released as 1.21.1

codefromthecrypt commented:
ETA tomorrow!

codefromthecrypt commented:
expect an update tomorrow pacific AM. Just polishing tests

Commit:
Adds Span.timestamp, duration to json and thrift representations

With this in place, instrumentation can send timestamp and duration
explicitly, which facilitates local spans or other spans who don't log
RPC annotations.

codefromthecrypt commented:
ok. this is finally complete. Once merged, I'll raise a pull request for duration query.
